File: backend/app/api/v1/leads.py
"""线索池 API。"""
from typing import Annotated, Optional

from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app.api.deps import PermissionChecker, get_current_active_user
from app.database import get_db
from app.models.user import User
from app.schemas.lead import (
    DuplicateCheckOut,
    LeadAssignRequest,
    LeadBatchAssignRequest,
    LeadBatchAssignResult,
    LeadConvertOut,
    LeadConvertRequest,
    LeadCreate,
    LeadDetailOut,
    LeadFollowUpCreate,
    LeadFollowUpOut,
    LeadListOut,
    LeadLostRequest,
    LeadOut,
    LeadQuotaOut,
    LeadReturnRequest,
    LeadStatsOut,
    LeadTransferRequest,
    LeadUpdate,
)
from app.services import lead as lead_service

router = APIRouter(prefix="/leads", tags=["线索池"])


@router.get("/stats", response_model=LeadStatsOut, summary="线索统计")
def stats(
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(PermissionChecker(["lead:view"]))],
) -> LeadStatsOut:
    return LeadStatsOut(**lead_service.lead_stats(db, current_user))


@router.get("/quota", response_model=LeadQuotaOut, summary="我的公海抢领额度")
def quota(
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(PermissionChecker(["lead:view"]))],
) -> LeadQuotaOut:
    return LeadQuotaOut(**lead_service.get_lead_quota(db, current_user))


@router.get("/duplicates", response_model=DuplicateCheckOut, summary="重复检测")
def check_duplicates(
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(PermissionChecker(["lead:view"]))],
    phone: Optional[str] = None,
    company_name: Optional[str] = None,
    credit_code: Optional[str] = None,
    company_domain: Optional[str] = None,
) -> DuplicateCheckOut:
    dups = lead_service.find_duplicates(
        db,
        phone=phone,
        company_name=company_name,
        credit_code=credit_code,
        company_domain=company_domain,
    )

    def _out(items):
        return [LeadOut.model_validate(lead_service.enrich_lead(db, x)) for x in items]

    hard = bool(dups["by_phone"] or dups["by_credit"])
    soft = bool(dups["by_company"] or dups["by_domain"])
    return DuplicateCheckOut(
        has_duplicate=hard or soft,
        is_hard_duplicate=hard,
        by_phone=_out(dups["by_phone"]),
        by_company=_out(dups["by_company"]),
        by_credit=_out(dups["by_credit"]),
        by_domain=_out(dups["by_domain"]),
    )


@router.get("", response_model=LeadListOut, summary="线索列表")
def list_leads(
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(PermissionChecker(["lead:view"]))],
    status: Optional[str] = None,
    keyword: Optional[str] = None,
    pool: Optional[str] = Query(None, description="mine/created/public/all"),
    business_type: Optional[str] = None,
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
) -> LeadListOut:
    total, items = lead_service.list_leads(
        db,
        current_user,
        status=status,
        keyword=keyword,
        pool=pool,
        business_type=business_type,
        page=page,
        page_size=page_size,
    )
    return LeadListOut(total=total, items=[LeadOut.model_validate(x) for x in items])


@router.post("", response_model=LeadOut, summary="录入线索")
def create_lead(
    payload: LeadCreate,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(PermissionChecker(["lead:view"], any_of=False))],
    force: bool = Query(False, description="手机号重复时强制创建"),
) -> LeadOut:
    # 全员有 lead:view 即可录入（普通员工种子会补上）
    lead = lead_service.create_lead(db, current_user, payload, force=force)
    return LeadOut.model_validate(lead)


@router.get("/{lead_id}", response_model=LeadDetailOut, summary="线索详情")
def get_lead(
    lead_id: int,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(PermissionChecker(["lead:view"]))],
) -> LeadDetailOut:
    lead = lead_service.get_lead_detail(db, current_user, lead_id)
    return LeadDetailOut.model_validate(lead)


@router.patch("/{lead_id}", response_model=LeadOut, summary="编辑线索")
def update_lead(
    lead_id: int,
    payload: LeadUpdate,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(PermissionChecker(["lead:view"]))],
) -> LeadOut:
    lead = lead_service.update_lead(db, current_user, lead_id, payload)
    return LeadOut.model_validate(lead)


@router.post("/{lead_id}/assign", response_model=LeadOut, summary="手动分配")
def assign_lead(
    lead_id: int,
    payload: LeadAssignRequest,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(PermissionChecker(["lead:manage"]))],
) -> LeadOut:
    lead = lead_service.assign_lead(db, current_user, lead_id, payload)
    return LeadOut.model_validate(lead)


@router.post("/batch-assign", response_model=LeadBatchAssignResult, summary="批量分配待分配线索")
def batch_assign_leads(
    payload: LeadBatchAssignRequest,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(PermissionChecker(["lead:manage"]))],
) -> LeadBatchAssignResult:
    result = lead_service.batch_assign_leads(
        db,
        current_user,
        lead_ids=payload.lead_ids,
        owner_ids=payload.owner_ids,
        method=payload.method,
        assignments=[x.model_dump() for x in payload.assignments],
        reason=payload.reason,
    )
    return LeadBatchAssignResult(**result)


@router.post("/{lead_id}/claim", response_model=LeadOut, summary="公海领取（兼容旧入口，正式流程请用分配）")
def claim_lead(
    lead_id: int,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_active_user)],
) -> LeadOut:
    lead = lead_service.claim_lead(db, current_user, lead_id)
    return LeadOut.model_validate(lead)


@router.post("/{lead_id}/transfer", response_model=LeadOut, summary="流转给他人")
def transfer_lead(
    lead_id: int,
    payload: LeadTransferRequest,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(PermissionChecker(["lead:view"]))],
) -> LeadOut:
    lead = lead_service.transfer_lead(db, current_user, lead_id, payload)
    return LeadOut.model_validate(lead)


@router.post("/{lead_id}/return", response_model=LeadOut, summary="退回公海")
def return_lead(
    lead_id: int,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(PermissionChecker(["lead:view"]))],
    payload: Optional[LeadReturnRequest] = None,
    reason: Optional[str] = Query(None, description="兼容旧版 query 参数"),
) -> LeadOut:
    body = payload or LeadReturnRequest()
    lead = lead_service.return_to_pool(
        db,
        current_user,
        lead_id,
        body.reason or reason,
        reason_type=body.reason_type,
    )
    return LeadOut.model_validate(lead)


@router.post("/{lead_id}/follow-ups", response_model=LeadFollowUpOut, summary="写跟进")
def create_follow_up(
    lead_id: int,
    payload: LeadFollowUpCreate,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(PermissionChecker(["lead:view"]))],
) -> LeadFollowUpOut:
    fu = lead_service.add_follow_up(db, current_user, lead_id, payload)
    return LeadFollowUpOut.model_validate(fu)


@router.post("/{lead_id}/convert", response_model=LeadConvertOut, summary="转化为客户与商机")
def convert_lead(
    lead_id: int,
    payload: LeadConvertRequest,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(PermissionChecker(["lead:view"]))],
) -> LeadConvertOut:
    result = lead_service.convert_lead(db, current_user, lead_id, payload)
    return LeadConvertOut(
        lead=LeadOut.model_validate(result["lead"]),
        customer_id=result["customer_id"],
        opportunity_id=result["opportunity_id"],
    )


@router.post("/{lead_id}/lost", response_model=LeadOut, summary="标记流失")
def lost_lead(
    lead_id: int,
    payload: LeadLostRequest,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(PermissionChecker(["lead:view"]))],
) -> LeadOut:
    lead = lead_service.mark_lost(db, current_user, lead_id, payload)
    return LeadOut.model_validate(lead)

File: frontend/src/api/leads.ts
/**
 * 线索池 API
 */
import request from './request'

export type LeadStatus =
  | 'pending_assign'
  | 'assigned'
  | 'following'
  | 'converted'
  | 'returned'
  | 'lost'

export interface Lead {
  id: number
  name: string
  company_name?: string | null
  credit_code?: string | null
  company_domain?: string | null
  phone?: string | null
  email?: string | null
  region?: string | null
  source: string
  source_detail?: string | null
  need_desc?: string | null
  budget?: number | null
  expected_deal_at?: string | null
  business_type?: string | null
  status: LeadStatus | string
  owner_id?: number | null
  creator_id?: number | null
  department_id?: number | null
  protect_until?: string | null
  assigned_at?: string | null
  last_followed_at?: string | null
  converted_customer_id?: number | null
  converted_opportunity_id?: number | null
  converted_at?: string | null
  lost_reason?: string | null
  lost_at?: string | null
  remark?: string | null
  created_at: string
  updated_at: string
  owner_name?: string | null
  creator_name?: string | null
  is_protected?: boolean
}

export interface LeadFollowUp {
  id: number
  lead_id: number
  user_id: number
  follow_at: string
  method: string
  content: string
  customer_feedback?: string | null
  result: string
  next_follow_at?: string | null
  created_at: string
}

export interface LeadLog {
  id: number
  lead_id: number
  user_id?: number | null
  username?: string | null
  action: string
  detail?: string | null
  created_at: string
}

export interface LeadDetail extends Lead {
  follow_ups: LeadFollowUp[]
  logs: LeadLog[]
}

export interface LeadStats {
  total: number
  pending_assign: number
  assigned: number
  following: number
  converted: number
  returned: number
  lost: number
  public_pool: number
  today_created?: number
  today_assigned?: number
  following_mine?: number
  protect_expiring?: number
  converted_month?: number
  mine?: number
  created?: number
  created_pending_assign?: number
  created_assigned?: number
  created_following?: number
  created_converted?: number
}

export interface LeadQuota {
  daily_claimed: number
  daily_limit: number
  protected_count: number
  protect_limit: number
  protect_days: number
  cooldown_hours: number
  can_claim: boolean
  block_reason?: string | null
}

export const LEAD_STATUS_LABEL: Record<string, string> = {
  pending_assign: '待分配',
  assigned: '已分配',
  following: '跟进中',
  converted: '已转化',
  returned: '已退回',
  lost: '已流失',
}

export const LEAD_SOURCE_OPTIONS = [
  { value: 'manual', label: '手动录入' },
  { value: 'import', label: '批量导入' },
  { value: 'external', label: '外部筛选' },
  { value: 'website', label: '官网' },
  { value: 'ad', label: '广告投放' },
  { value: 'event', label: '展会/活动' },
  { value: 'referral', label: '转介绍' },
  { value: 'im', label: '飞书/企微' },
  { value: 'other', label: '其他' },
]

export const LEAD_RETURN_REASON_OPTIONS = [
  { value: 'no_need', label: '客户无需求' },
  { value: 'unreachable', label: '联系不上' },
  { value: 'competitor', label: '竞品已签约' },
  { value: 'budget', label: '预算不足' },
  { value: 'other', label: '其他' },
]

export function fetchLeadStats() {
  return request.get<LeadStats>('/leads/stats')
}

export function fetchLeadQuota() {
  return request.get<LeadQuota>('/leads/quota')
}

export function checkLeadDuplicates(params: {
  phone?: string
  company_name?: string
  credit_code?: string
  company_domain?: string
}) {
  return request.get<{
    has_duplicate: boolean
    is_hard_duplicate: boolean
    by_phone: Lead[]
    by_company: Lead[]
    by_credit: Lead[]
    by_domain: Lead[]
  }>('/leads/duplicates', { params })
}

export function fetchLeads(params: {
  status?: string
  keyword?: string
  pool?: string
  business_type?: string
  page?: number
  page_size?: number
}) {
  return request.get<{ total: number; items: Lead[] }>('/leads', { params })
}

export function fetchLeadDetail(id: number) {
  return request.get<LeadDetail>(`/leads/${id}`)
}

export function createLead(data: Partial<Lead>, force = false) {
  return request.post<Lead>('/leads', data, { params: { force } })
}

export function updateLead(id: number, data: Partial<Lead>) {
  return request.patch<Lead>(`/leads/${id}`, data)
}

export function claimLead(id: number) {
  return request.post<Lead>(`/leads/${id}/claim`)
}

export function assignLead(id: number, owner_id: number, remark?: string) {
  return request.post<Lead>(`/leads/${id}/assign`, { owner_id, remark })
}

export function batchAssignLeads(data: {
  lead_ids: number[]
  owner_ids?: number[]
  method?: 'average' | 'manual'
  assignments?: { lead_id: number; owner_id: number }[]
  reason?: string
}) {
  return request.post<{
    success_count: number
    failed_count: number
    success: { lead_id: number; owner_id?: number; owner_name?: string }[]
    failed: { lead_id: number; reason?: string }[]
  }>('/leads/batch-assign', data)
}

export function transferLead(id: number, owner_id: number, reason?: string) {
  return request.post<Lead>(`/leads/${id}/transfer`, { owner_id, reason })
}

export function returnLead(
  id: number,
  payload?: { reason_type?: string; reason?: string } | string,
) {
  if (typeof payload === 'string' || payload === undefined) {
    return request.post<Lead>(`/leads/${id}/return`, null, {
      params: payload ? { reason: payload } : undefined,
    })
  }
  return request.post<Lead>(`/leads/${id}/return`, payload)
}

export function convertLead(
  id: number,
  data?: {
    customer_name?: string
    remark?: string
    opportunity_title?: string
    opportunity_stage?: string
    expected_amount?: number
    business_type?: string
    requirement_summary?: string
  },
) {
  return request.post<{
    lead: Lead
    customer_id: number
    opportunity_id: number
  }>(`/leads/${id}/convert`, data || {})
}

export function markLeadLost(id: number, reason: string) {
  return request.post<Lead>(`/leads/${id}/lost`, { reason })
}

export function createFollowUp(
  id: number,
  data: {
    method?: string
    content: string
    customer_feedback?: string
    result?: string
    next_follow_at?: string
  },
) {
  return request.post<LeadFollowUp>(`/leads/${id}/follow-ups`, data)
}

export interface LeadImportPreviewRow {
  row_no: number
  company_name: string
  phone: string
  name?: string | null
  credit_code?: string | null
  company_domain?: string | null
  business_type: string
  business_type_label: string
  need_desc?: string | null
  remark?: string | null
  status: 'ok' | 'soft' | 'hard' | 'error' | string
  message: string
  can_import: boolean
  force_required: boolean
}

export interface LeadImportPreview {
  total: number
  ok_count: number
  soft_count: number
  hard_count: number
  error_count: number
  rows: LeadImportPreviewRow[]
}

export interface LeadImportConfirmItem {
  row_no: number
  ok: boolean
  lead_id?: number | null
  message: string
}

export function downloadLeadImportTemplate() {
  return request.get<Blob>('/leads/import/template', { responseType: 'blob' })
}

export function previewLeadImport(file: File) {
  const form = new FormData()
  form.append('file', file)
  return request.post<LeadImportPreview>('/leads/import/preview', form, {
    headers: { 'Content-Type': 'multipart/form-data' },
    timeout: 60000,
  })
}

export function confirmLeadImport(data: {
  rows: Array<{
    row_no: number
    company_name: string
    phone: string
    name?: string | null
    credit_code?: string | null
    company_domain?: string | null
    business_type: string
    need_desc?: string | null
    remark?: string | null
    source?: string | null
    force?: boolean
  }>
  self_follow?: boolean | null
}) {
  return request.post<{
    success_count: number
    failed_count: number
    skipped_count: number
    items: LeadImportConfirmItem[]
  }>('/leads/import/confirm', data, { timeout: 120000 })
}
